### ISSUE TITLE
How can I detect shutdown signals from inside my workers?

### ISSUE THREAD

**Issue by dmagliola** [mperham/sidekiq#2364]
I'm doing deferred writing to my DB using Redis and Sidekiq. Basically writing records one by one is very slow, it's better to write them to Redis, and then every now and then do the inserts in batches.

For this, I have a number of Sidekiq jobs that essentially do:

    loop do
      rows = pop 100 rows from redis
      break if rows.empty?
      insert rows into DB all in one query
    end

Now, each iteration of this loop runs *very* quickly, but the job in its entirety may take minutes to run. And I have only have 8 seconds to shutdown, which I probably won't manage, and there are good odds I'll lose those 100 rows.

I'd like to have a "break if Sidekiq decided it's exiting soon" at the beginning or end of the loop, which practically guarantees I'll exit gracefully within 8 seconds.

Is there a flag somewhere I could look at to find this out, or some other best practice you'd recommend for this case?

Thank you

**Comment by mperham** [comment]
Are you using Heroku?  If not, extend your shutdown timeout to 30 seconds with `-t 30`.

You've built a long-running serial job.  I advise you to avoid this pattern.  Parallelize your work by creating a job for each 100 rows rather than having a single job with a `loop`.  With this change each job won't run long and you no longer care about shutdown.

**Comment by dmagliola** [comment]
I am writing inside Heroku...

As for the job per 100 rows, are you basically suggesting this?

    #no loop
    rows = pop 100 rows from redis
    break if rows.empty?
    insert rows into DB all in one query
    ThisJob.perform_now

So that it runs once and reschedules itself, unless there wasn't anything to do? Is this the most efficient way?

I saw through the code that the Manager has a stopped? attribute that seems like it'd do what I want, but I'm not exactly sure how to access it from inside my workers.

**Comment by mperham** [comment]
That's one way (although you want perform_later, not perform_now).  Or do it in parallel: `(llen / 100).times { ThisJob.perform_later }`

**Comment by mperham** [comment]
Short answer to your original question: I don't and won't expose the shutdown flag to workers because it enables anti-patterns like long-running serial jobs.

**Comment by dmagliola** [comment]
Fair enough. Thank you for always replying so quickly!!!

**Comment by dmagliola** [comment]
Actually... Sorry.... Looking at how I'd have to do this in code... That would imply a lot of very ugly hacks in my code to get this to work. It'd make it much, much harder to follow and understand. 

The gist of it is, kind of, the method that is currently looping, and which knows whether a new job needs to be scheduled doesn't know *which* job to schedule. I have multiple Worker classes dealing with this for multiple different models (and fields in some cases), all very similar but slightly different, so I'd need to pass around lots of parameters that pretty much make no sense.

I also *really* don't want to parallellize this, since dealing with all those batches of 100 records at once would completely grind my DB to a halt. That's precisely the opposite of what i'm trying to achieve, I don't care how long this takes overall as long as it gets done eventually, and I want it to let the rest of the system work happily in the meantime.

Why is the long running serial job an antipattern? (and any chance I can get access to that Manager?)

**Comment by mperham** [comment]
Sidekiq is designed to run lots of little jobs in parallel.  It is not appropriate for every type of work.  If you want to perform bulk loads in serial, I'd suggest using a cron job that runs every N minutes and loads whatever has accumulated.

**Comment by dmagliola** [comment]
Ok, sounds good. Thanks!

**Comment by mperham** [comment]
And for the record, I don't believe the workers have any "easy" access to the Manager instance.  There are hacks possible (e.g. ObjectSpace) but it's not designed to be a public API.
